packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,50 +3,197 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/painting.dart';
+import 'package:flutter/services.dart';
 
-/// A leaf node in the focus tree that can receive focus.
+import 'binding.dart';
+import 'focus_traversal.dart';
+import 'framework.dart';
+
+/// Signature of a callback used by [Focus.onKey] and [FocusScope.onKey]
+/// to receive key events.
+///
+/// The [node] is the node that received the event.
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
+
+/// An object that can be used by a stateful widget to obtain the keyboard focus
+/// and to handle keyboard events.
+///
+/// {@template flutter.widgets.focus_manager.focus.advancedWarning}
+/// **If you are not implementing a widget which handles focused input directly
+/// (e.g. implementing your own text field or other widget primitive), or
+/// keeping [FocusNode] references so that you can control the focus from
+/// another part of the application, you typically will not need to manage a
+/// [FocusNode] directly, as it will be handled by the [Focus] and [FocusScope]
+/// widgets or by the widget primitives themselves. Please see those widgets for
+/// typical usage.**
+/// {@endtemplate}
 ///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
+/// [FocusNode]s are persistent objects which form a "focus tree" that is a
+/// representation of the widgets in the hierarchy which are interested in
+/// focus. They must be managed like other persistent state, which is typically
+/// done by a [StatefulWidget]. A stateful widget that manages a focus node must
+/// call [dispose] from its [State.dispose] method.
 ///
-/// To request focus, find the [FocusScopeNode] for the current [BuildContext]
-/// and call the [FocusScopeNode.requestFocus] method:
+/// ## Hierarchy Management
 ///
-/// ```dart
-/// FocusScope.of(context).requestFocus(focusNode);
-/// ```
+/// Because [FocusNode]s form a sparse representation of the widget tree, they
+/// must be updated whenever the widget tree is rebuilt. This is done by calling
+/// [reparentIfNeeded], usually from a [State.build] method of the widget which
+/// represents the focused region, so that the size and location of the focused
+/// area are correctly represented in its context.
```

Comment:
someone reading this who knows how flutter works may get concerned at that point, since size and location information is not computed at the time that build is called. Maybe:

```
..., so that the [BuildContext] to which the [FocusNode] is assigned can be tracked (that context is
used to obtain the [RenderObject], from which the geometry of focus rings can be determined).
```

Reply:
(btw, take any of my suggestions for actual text such as here as mere writing prompts, not as graven tablets that must be used verbatim.)

Reply:
:-) sometimes i leave a drive-by comment on a PR and years later I find my two-second half-thought has been baked into the docs, so, you know, just making sure...